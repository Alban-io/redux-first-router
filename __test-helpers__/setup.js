import createHistory from 'history/createMemoryHistory'
import connectRoutes from '../src/connectRoutes'


export default (path = '/', options = { title: 'title', location: 'location' }) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: [path],
    initialIndex: 0,
    keyLength: 6,
  })

  return connectRoutes(history, routesMap, options)
}
