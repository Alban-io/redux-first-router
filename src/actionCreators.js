export const INIT = '@@address-bar/INIT' //used internally by middleware and then action is re-written
export const NOT_FOUND = '@@address-bar/NOT_FOUND'


/** EXPORTED INIT ACTION CREATORS: 
 *  for use when ready key is not provided and 
 *  manual control of initialization is desired
*/

export function init(pathname) {
  return {
    type: INIT,
    payload: {
      pathname,
    }
  }
}

export function initThunk() {
  return (dispatch, getState) => {
    let {pathname} = getState().location
    return dispatch(init(pathname))
  }
}


/** PRIMARY/SOLE NAVIGATION ACTION CREATOR 
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'