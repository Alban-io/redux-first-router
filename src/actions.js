export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

/** PRIMARY/SOLE NAVIGATION ACTION CREATOR
 *  note: on web most of the time you want to use a link component that embeds an `<a />` element
 *  to the DOM for SEO purposes. See our small `<Link />`component:
 *  https://github.com/celebvidy/pure-redux-router-link
*/

export { go } from './connectTypes'
