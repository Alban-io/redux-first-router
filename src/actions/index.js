export { default as redirect } from './redirect'
export { default as notFound } from './notFound'
export { default as addRoutes } from './addRoutes'
export { default as changeBasename } from './changeBasename'

export {
  push,
  replace,
  setState,
  jump,
  back,
  next,
  reset
} from './history'
