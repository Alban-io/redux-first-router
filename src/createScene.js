// @flow
import type { RoutesMapInput, CreateActionsOptions } from './flow-types'
import notFound from './action-creators/notFound'
import { NOT_FOUND } from './index'
import isFSRA from './utils/isFSRA'
import isNotFound from './utils/isNotFound'

export default (r: RoutesMapInput, opts: CreateActionsOptions = {}) => {
  const { scene: sc, basename: bn, logExports } = opts

  const scene = sc || ''
  const prefix = scene ? `${scene}/` : ''
  const keys = Object.keys(r)

  const result = keys.reduce((result, t) => {
    const { types, actions, routes } = result

    const t2 = `${prefix}${t}`
    const tc = `${prefix}${t}_COMPLETE`
    const te = `${prefix}${t}_ERROR`

    const route = routes[t2] = routeToObject(r[t], t2)
    const tClean = route.scene ? t2.replace(`${route.scene}/`, '') : t // strip the scene so keys/exports are un-prefixed
    const name = camelCase(tClean)

    types[tClean] = t2
    types[`${tClean}_COMPLETE`] = tc
    types[`${tClean}_ERROR`] = te

    // allow for creating custom action creators (whose names are an array assigned to route.action)
    if (Array.isArray(route.action)) {
      const key = route.action[0]
      actions[name] = makeAC(route, t2, key, bn) // the first name in the array becomes the primary action creator

      // all are tacked on like name.complete, name.error
      route.action.forEach((key: string) => {
        actions[name][key] = makeAC(route, t2, key, bn)
      })
    }
    else {
      actions[name] = makeAC(route, t2, 'action', bn)
    }

    actions[name].complete = makeAC(route, tc, 'complete', bn)
    actions[name].error = makeAC(route, te, 'error', bn)

    return result
  }, { types: {}, actions: {}, routes: {} })

  const { types, actions } = result

  // insure @@rudy/NOT_FOUND keys/exports are also un-prefixed, eg: NOT_FOUND, notFound, etc
  if (types[NOT_FOUND]) {
    types.NOT_FOUND = types[NOT_FOUND]
    types.NOT_FOUND_COMPLETE = types[`${NOT_FOUND}_COMPLETE`]
    types.NOT_FOUND_ERROR = types[`${NOT_FOUND}_ERROR`]
    delete types[NOT_FOUND]
    delete types[`${NOT_FOUND}_COMPLETE`]
    delete types[`${NOT_FOUND}_ERROR`]

    actions.notFound = actions.rudyNotFound
    delete actions.rudyNotFound
  }

  if (logExports && /development|test/.test(process.env.NODE_ENV)) {
    result.exportString = logExportString(types, actions, result.routes, opts)
  }

  return result
}

const routeToObject = (route, type) => {
  const r = typeof route === 'function'
    ? { thunk: route }
    : typeof route === 'string'
      ? { path: route }
      : route

  r.scene = getScene(type)

  // set default path for NOT_FOUND actions if necessary
  if (!r.path && isNotFound(type)) {
    r.path = r.scene ? `/${r.scene.toLowerCase()}/not-found` : '/not-found'
  }

  return r
}

const getScene = (type: string) => {
  const i = type.lastIndexOf('/')
  return type.substr(0, i).replace(/\/?@@rudy/, '')
}

const makeAC = (
  route: Object, // primary type
  type: string,
  key: ?string,
  basename: ?string
) => {
  const ac = typeof route[key] === 'function' ? route[key] : null // look for action creators on route

  // `info` arg contains 'isThunk' or optional `path` for `notFound` action creators
  const defaultCreator = (arg: Object | Function, info: ?string) => {
    // optionally handle action creators that return functions (aka `thunk`)
    if (typeof arg === 'function') {
      const thunk: Function = arg
      return (...args: Array<any>) => defaultCreator(thunk(...args, type), 'isThunk')
    }

    // do nothing if a `thunk` returned nothing (i.e. manually used `dispatch`)
    if (info === 'isThunk' && arg === undefined) return

    // for good measure honor promises (`dispatch` will have manually been used)
    if (info === 'isThunk' && arg && arg.then) return arg

    // use built-in `notFound` action creator if `NOT_FOUND` type
    const t = (arg && arg.type) || type
    if (isNotFound(t)) {
      const notFoundPath = info === 'isThunk' ? null : info
      return notFound(arg, notFoundPath, basename, t)
    }

    // handle error action creator
    if (key === 'error') return handleError(arg, t, basename)

    // the default behavior of transforming an `arg` object into an action with its type
    if (isFSRA(arg)) return { type, ...arg, location: { basename, ...arg.location } }

    // if no `payload`, `query`, etc, treat arg as a `payload` for convenience
    return { type, payload: arg || {}, location: { basename } }
  }

  // optionally allow custom action creators
  if (ac) {
    return (...args: Array<any>) => defaultCreator(ac(...args, type))
  }

  // primary use case: generate an action creator (will only trigger last lines of `defaultCreator`)
  return defaultCreator
}


const handleError = (o: Object, type: string, basename: ?string) =>
  o && o.error
    ? { type, ...o, location: { basename, ...o.location } }
    : { type, error: o, location: { basename } }


const logExportString = (types, actions, routes, options) => {
  const opts = { ...options }
  opts.scene = getScene(Object.keys(routes)[0])
  delete opts.logExports

  const optsString = JSON.stringify(opts)
    .replace(/"scene":/, 'scene: ')
    .replace(/"basename":/, 'basename: ')
    .replace(/"/g, "'")
    .replace('{', '{ ')
    .replace('}', ' }')
    .replace(/,/g, ', ')

  let t = ''
  for (const type in types) t += `\n\t${type},`

  let a = ''
  for (const action in actions) a += `\n\t${action},`

  // destructure createActions()
  let exports = `const { types, actions } = createScene(routes, ${optsString})`
  exports += '\n\nconst { ' + t.slice(0, -1) + '\n} = types'
  exports += '\n\nconst { ' + a.slice(0, -1) + '\n} = actions'

  // types exports
  exports += '\n\nexport {' + t
  exports = exports.slice(0, -1) + '\n}'

  // actions exports
  exports += '\n\nexport {' + a
  exports = exports.slice(0, -1) + '\n}'

  if (process.env.NODE_ENV !== 'test') console.log(exports)
  return exports
}


const camelCase = (type: string) => {
  const matches = type.match(wordPattern)

  if (!Array.isArray(matches)) throw new Error(`[rudy] invalid action type: ${type}`)

  return matches.reduce((camelCased, word, index) =>
    camelCased + (index === 0
      ? word.toLowerCase()
      : word.charAt(0).toUpperCase() + word.substring(1).toLowerCase())
  , '')
}

const wordPattern = /[A-Z\xc0-\xd6\xd8-\xde]?[a-z\xdf-\xf6\xf8-\xff]+(?:['’](?:d|ll|m|re|s|t|ve))?(?=[\xac\xb1\xd7\xf7\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xbf\u2000-\u206f \t\x0b\f\xa0\ufeff\n\r\u2028\u2029\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000]|[A-Z\xc0-\xd6\xd8-\xde]|$)|(?:[A-Z\xc0-\xd6\xd8-\xde]|[^\ud800-\udfff\xac\xb1\xd7\xf7\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xbf\u2000-\u206f \t\x0b\f\xa0\ufeff\n\r\u2028\u2029\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000\d+\u2700-\u27bfa-z\xdf-\xf6\xf8-\xffA-Z\xc0-\xd6\xd8-\xde])+(?:['’](?:D|LL|M|RE|S|T|VE))?(?=[\xac\xb1\xd7\xf7\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xbf\u2000-\u206f \t\x0b\f\xa0\ufeff\n\r\u2028\u2029\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000]|[A-Z\xc0-\xd6\xd8-\xde](?:[a-z\xdf-\xf6\xf8-\xff]|[^\ud800-\udfff\xac\xb1\xd7\xf7\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xbf\u2000-\u206f \t\x0b\f\xa0\ufeff\n\r\u2028\u2029\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000\d+\u2700-\u27bfa-z\xdf-\xf6\xf8-\xffA-Z\xc0-\xd6\xd8-\xde])|$)|[A-Z\xc0-\xd6\xd8-\xde]?(?:[a-z\xdf-\xf6\xf8-\xff]|[^\ud800-\udfff\xac\xb1\xd7\xf7\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xbf\u2000-\u206f \t\x0b\f\xa0\ufeff\n\r\u2028\u2029\u1680\u180e\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200a\u202f\u205f\u3000\d+\u2700-\u27bfa-z\xdf-\xf6\xf8-\xffA-Z\xc0-\xd6\xd8-\xde])+(?:['’](?:d|ll|m|re|s|t|ve))?|[A-Z\xc0-\xd6\xd8-\xde]+(?:['’](?:D|LL|M|RE|S|T|VE))?|\d*(?:(?:1ST|2ND|3RD|(?![123])\dTH)\b)|\d*(?:(?:1st|2nd|3rd|(?![123])\dth)\b)|\d+|(?:[\u2700-\u27bf]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff])[\ufe0e\ufe0f]?(?:[\u0300-\u036f\ufe20-\ufe2f\u20d0-\u20ff]|\ud83c[\udffb-\udfff])?(?:\u200d(?:[^\ud800-\udfff]|(?:\ud83c[\udde6-\uddff]){2}|[\ud800-\udbff][\udc00-\udfff])[\ufe0e\ufe0f]?(?:[\u0300-\u036f\ufe20-\ufe2f\u20d0-\u20ff]|\ud83c[\udffb-\udfff])?)*/g
