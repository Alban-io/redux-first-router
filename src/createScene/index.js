// @flow
import type { RoutesMapInput, CreateActionsOptions } from '../flow-types'
import { NOT_FOUND } from '../types'

import {
  camelCase,
  logExports,
  makeActionCreator,
  formatRoute
} from './utils'

export default (routesMap: RoutesMapInput, opts: CreateActionsOptions = {}) => {
  const { scene: sc, basename: bn, formatRoute: format, logExports: log } = opts

  const scene = sc || ''
  const prefix = scene ? `${scene}/` : ''
  const keys = Object.keys(routesMap)

  const result = keys.reduce((result, t) => {
    const { types, actions, routes } = result

    const t2 = `${prefix}${t}`
    const tc = `${prefix}${t}_COMPLETE`
    const te = `${prefix}${t}_ERROR`

    routes[t2] = formatRoute(routesMap[t], t2, routesMap, format)

    const route = routes[t2]
    const tClean = route.scene ? t2.replace(`${route.scene}/`, '') : t // strip the scene so types will be un-prefixed
    const action = camelCase(tClean)

    types[tClean] = t2
    types[`${tClean}_COMPLETE`] = tc
    types[`${tClean}_ERROR`] = te

    // allow for creating custom action creators (whose names are an array assigned to route.action)
    if (Array.isArray(route.action)) {
      const key = route.action[0]
      actions[action] = makeActionCreator(route, t2, key, bn) // the first action in the array becomes the primary action creator

      // all are tacked on like action.complete, action.error
      route.action.forEach((key: string) => {
        actions[action][key] = makeActionCreator(route, t2, key, bn)
      })
    }
    else {
      actions[action] = makeActionCreator(route, t2, 'action', bn)
    }

    actions[action].complete = makeActionCreator(route, tc, 'complete', bn)
    actions[action].error = makeActionCreator(route, te, 'error', bn)

    return result
  }, { types: {}, actions: {}, routes: {} })

  const { types, actions } = result

  // insure @@rudy/NOT_FOUND routes are also un-prefixed, eg: NOT_FOUND, notFound, etc
  if (types[NOT_FOUND]) {
    types.NOT_FOUND = types[NOT_FOUND]
    types.NOT_FOUND_COMPLETE = types[`${NOT_FOUND}_COMPLETE`]
    types.NOT_FOUND_ERROR = types[`${NOT_FOUND}_ERROR`]
    delete types[NOT_FOUND]
    delete types[`${NOT_FOUND}_COMPLETE`]
    delete types[`${NOT_FOUND}_ERROR`]

    actions.notFound = actions.rudyNotFound
    delete actions.rudyNotFound
  }

  if (log && /^(development|test)$/.test(process.env.NODE_ENV)) {
    result.exportString = logExports(types, actions, result.routes, opts)
  }

  return result
}
