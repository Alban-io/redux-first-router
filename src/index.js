export {
  default as connectRoutes,
  go,
  back,
} from './connectRoutes'

export const NOT_FOUND = '@@pure-redux-router/NOT_FOUND'

export { default as actionToPath } from './pure-utils/actionToPath'
export { default as pathToAction } from './pure-utils/pathToAction'

export type {
  RouteString,
  RouteObject,
  Route,
  RoutesMap,
  Routes,
  RouteNames,
  Options,
  Params,
  Payload,
  LocationState,
  Location,
  Meta,
  Action,
  ReceivedAction,
  Listener,
  Listen,
  Push,
  GoBack,
  History,
  HistoryLocation,
  Document,
 } from './flow-types'

