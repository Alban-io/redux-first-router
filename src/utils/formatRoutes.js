// @flow
import { ADD_ROUTES, CHANGE_BASENAME, CLEAR_CACHE, CONFIRM, CALL_HISTORY } from '../types'
import { redirect } from '../actions'
import type { RoutesMap, RoutesMapInput } from '../flow-types'

import {
  addRoutes,
  changeBasename,
  clearCache,
  confirm,
  callHistory
} from '../pathlessRoutes'

export default (
  routes: RoutesMapInput,
  formatRoute: ?Function,
  isAddRoutes: boolean = false
): RoutesMap => {
  if (!isAddRoutes) {
    routes.NOT_FOUND = routes.NOT_FOUND || { path: '/not-found' }
  }

  routes[ADD_ROUTES] = routes[ADD_ROUTES] || { thunk: addRoutes }
  routes[CHANGE_BASENAME] = routes[CHANGE_BASENAME] || { thunk: changeBasename, dispatch: false }
  routes[CLEAR_CACHE] = routes[CLEAR_CACHE] || { thunk: clearCache }
  routes[CONFIRM] = routes[CONFIRM] || { thunk: confirm, dispatch: false }
  routes[CALL_HISTORY] = routes[CALL_HISTORY] || { thunk: callHistory, dispatch: false }

  for (const type in routes) {
    const route = format(routes[type], type, routes, formatRoute, isAddRoutes)

    route.type = type
    if (route.redirect) createRedirect(route, routes)
    routes[type] = route
  }

  return routes
}

export const format = (r, type, routes, formatRoute, isAddRoutes) => {
  const route = typeof r === 'string' ? { path: r } : r

  if (formatRoute) {
    return formatRoute(route, type, routes, isAddRoutes)
  }

  if (typeof route === 'function') {
    return { thunk: route }
  }

  return route
}

const createRedirect = (route, routes) => {
  const t = route.redirect
  const scenicType = `${route.scene}/${t}`
  const type = routes[scenicType] ? scenicType : t

  route.redirectBeforeEnter = ({ action, dispatch }) => {
    return dispatch(redirect({ ...action, type }, 301))
  }
}
