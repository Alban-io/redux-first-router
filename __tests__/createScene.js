import setup, { defaultRoutes, log, me } from '../__test-helpers__/rudySetup'
import { NOT_FOUND } from '../src'
import createScene from '../src/createScene'
import formatRoutes from '../src/utils/formatRoutes'

test('createScene', async () => {
  const tools = setup(undefined, undefined, { scene: 'SCENE', logExports: true })
  const { store, firstRoute, types, actions, exportString, routes } = tools
  let action = firstRoute()

  let res = await store.dispatch(action)
  expect(store.getState().location.type).toEqual('SCENE/FIRST')

  res = await store.dispatch(actions.second())
  console.log(actions.second())
  expect(store.getState().location.type).toEqual('SCENE/SECOND')

  action = actions.third((req, type) => ({ testReq: req.getTitle(), type }))
  res = await store.dispatch(action)
  expect(store.getState().location.type).toEqual('SCENE/THIRD')

  action = actions.fourth('baz')
  res = await store.dispatch(action)
  console.log(action)
  expect(store.getState().location.type).toEqual('SCENE/FOURTH')
  expect(res).toEqual({ type: 'SCENE/FOURTH_COMPLETE', payload: 'onComplete' })

  expect(store.getState().location.payload).toEqual({ foo: 'baz' })

  res = await store.dispatch(actions.second({ payload: { foo: 'bar' } }))
  expect(store.getState().location.type).toEqual('SCENE/SECOND')
  expect(store.getState().location.payload).toEqual({ foo: 'bar' })

  res = await store.dispatch(actions.third({ type: 'WRONG' }))
  expect(store.getState().location.type).toEqual('SCENE/SECOND')

  res = await store.dispatch(actions.third.error(new Error('fail')))
  console.log('RES', res)
  expect(store.getState().location.type).toEqual('SCENE/SECOND')
  expect(store.getState().location.error).toEqual(new Error('fail'))
  expect(store.getState().location.errorType).toEqual('SCENE/THIRD_ERROR')

  res = await store.dispatch(actions.third.complete({ foo: 'bar' }))
  expect(res.type).toEqual('SCENE/THIRD_COMPLETE')
  expect(res.payload).toEqual({ foo: 'bar' })

  res = await store.dispatch(actions.plain('bar'))
  expect(res.type).toEqual('SCENE/PLAIN')
  expect(res.payload).toEqual({ foo: 'bar' })

  log(store)

  expect(types).toMatchSnapshot()
  expect(actions).toMatchSnapshot()
  expect(exportString).toMatchSnapshot()
})

test('NOT_FOUND', async () => {
  const tools = setup(undefined, undefined, { scene: 'SCENE', logExports: true })
  const { store, firstRoute, types, actions, exportString, routes } = tools
  let action = firstRoute()

  console.log(routes)

  let res = await store.dispatch(action)
  expect(store.getState().location.type).toEqual('SCENE/FIRST')

  action = actions.notFound({ foo: 'bar' }, 'cat')
  res = await store.dispatch(action)

  expect(store.getState().location.type).toEqual('SCENE/@@rudy/NOT_FOUND')

  expect(action).toMatchSnapshot()
  expect(store.getState().location).toMatchSnapshot()
})


test('double createScene', async () => {
  const { routes: r } = createScene(defaultRoutes, { scene: 'scene' })
  const { types, actions, routes: r2, exportString } = createScene(r, { scene: 'double', logExports: true })
  const routes = formatRoutes(r2)

  const custom = { createScene: false, routesMap: routes }
  const { store, firstRoute, location } = setup('/first', undefined, custom)
  let action
  let res

  action = firstRoute()
  res = await store.dispatch(action)
  expect(res.type).toEqual('double/scene/FIRST')
  expect(location().type).toEqual('double/scene/FIRST')

  action = actions.second({ foo: 'bar' })
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.SECOND)
  expect(res.type).toEqual('double/scene/SECOND')
  expect(location().type).toEqual('double/scene/SECOND')
  expect(location().payload).toEqual({ foo: 'bar' })

  action = { type: types.THIRD }
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.THIRD_COMPLETE)
  expect(res.type).toEqual('double/scene/THIRD_COMPLETE')
  expect(location().type).toEqual('double/scene/THIRD')

  action = actions.fourth('bar')
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.FOURTH_COMPLETE)
  expect(res.type).toEqual('double/scene/FOURTH_COMPLETE')
  expect(location().type).toEqual('double/scene/FOURTH')
  expect(location().payload).toEqual({ foo: 'bar' })

  action = actions.notFound(() => ({ foo: 'bar' }))
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.NOT_FOUND)
  expect(res.type).toEqual('double/scene/@@rudy/NOT_FOUND')
  expect(location().type).toEqual('double/scene/@@rudy/NOT_FOUND')
  expect(location().payload).toEqual({ foo: 'bar' })
  expect(location().pathname).toEqual('/not-found-foo')

  action = actions.second.error({ foo: 'bar' })
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.SECOND_ERROR)
  expect(res.type).toEqual('double/scene/SECOND_ERROR')
  expect(location().error).toEqual({ foo: 'bar', bla: 'boo' })
  expect(store.getState().title).toEqual('double/scene/SECOND_ERROR')

  action = actions.third.complete(() => ({ foo: 'bar' }))
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.THIRD_COMPLETE)
  expect(res.type).toEqual('double/scene/THIRD_COMPLETE')
  expect(res.payload).toEqual({ foo: 'bar' })
  expect(store.getState().title).toEqual('double/scene/THIRD_COMPLETE')

  expect(location().type).toEqual('double/scene/@@rudy/NOT_FOUND') // check that previous actions didn't change routes

  action = actions.third.customCreator('bar')
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.THIRD_COMPLETE)
  expect(res.type).toEqual('double/scene/THIRD_COMPLETE')
  expect(location().type).toEqual('double/scene/THIRD')
  expect(location().payload).toEqual({ foo: 'bar' })

  action = actions.notFound.complete('bar')
  res = await store.dispatch(action)
  expect(res.type).toEqual(types.NOT_FOUND_COMPLETE)
  expect(res.type).toEqual('double/scene/@@rudy/NOT_FOUND_COMPLETE')

  // location state equals previous state, since COMPLETE doesn't change routes
  expect(location().type).toEqual('double/scene/THIRD')
  expect(location().payload).toEqual({ foo: 'bar' })

  expect(types).toMatchSnapshot()
  expect(actions).toMatchSnapshot()
  expect(routes).toMatchSnapshot()
  expect(exportString).toMatchSnapshot()
})
