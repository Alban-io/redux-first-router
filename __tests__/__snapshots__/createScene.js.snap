// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NOT_FOUND 1`] = `
Object {
  "location": Object {
    "basename": undefined,
    "url": "/cat",
  },
  "params": Object {
    "foo": "bar",
  },
  "type": "SCENE/@@rudy/NOT_FOUND",
}
`;

exports[`NOT_FOUND 2`] = `
Object {
  "basename": "",
  "entries": Array [
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/first",
      "search": "",
      "state": Object {},
      "url": "/first",
    },
    Object {
      "hash": "",
      "key": "345678",
      "pathname": "/cat",
      "search": "",
      "state": Object {},
      "url": "/cat",
    },
  ],
  "hasSSR": false,
  "hash": "",
  "index": 1,
  "kind": "push",
  "length": 2,
  "params": Object {
    "foo": "bar",
  },
  "pathname": "/cat",
  "prev": Object {
    "basename": "",
    "entries": Array [
      Object {
        "hash": "",
        "key": "345678",
        "pathname": "/first",
        "search": "",
        "state": Object {},
        "url": "/first",
      },
    ],
    "hasSSR": false,
    "hash": "",
    "index": 0,
    "kind": "load",
    "length": 1,
    "params": Object {},
    "pathname": "/first",
    "query": Object {},
    "scene": "SCENE",
    "search": "",
    "state": Object {},
    "status": 200,
    "type": "SCENE/FIRST",
    "url": "/first",
  },
  "query": Object {},
  "scene": "SCENE",
  "search": "",
  "state": Object {},
  "status": 404,
  "type": "SCENE/@@rudy/NOT_FOUND",
  "url": "/cat",
}
`;

exports[`createScene 1`] = `
Object {
  "FIRST": "SCENE/FIRST",
  "FIRST_COMPLETE": "SCENE/FIRST_COMPLETE",
  "FIRST_ERROR": "SCENE/FIRST_ERROR",
  "FOURTH": "SCENE/FOURTH",
  "FOURTH_COMPLETE": "SCENE/FOURTH_COMPLETE",
  "FOURTH_ERROR": "SCENE/FOURTH_ERROR",
  "NOT_FOUND": "SCENE/@@rudy/NOT_FOUND",
  "NOT_FOUND_COMPLETE": "SCENE/@@rudy/NOT_FOUND_COMPLETE",
  "NOT_FOUND_ERROR": "SCENE/@@rudy/NOT_FOUND_ERROR",
  "PLAIN": "SCENE/PLAIN",
  "PLAIN_COMPLETE": "SCENE/PLAIN_COMPLETE",
  "PLAIN_ERROR": "SCENE/PLAIN_ERROR",
  "SECOND": "SCENE/SECOND",
  "SECOND_COMPLETE": "SCENE/SECOND_COMPLETE",
  "SECOND_ERROR": "SCENE/SECOND_ERROR",
  "THIRD": "SCENE/THIRD",
  "THIRD_COMPLETE": "SCENE/THIRD_COMPLETE",
  "THIRD_ERROR": "SCENE/THIRD_ERROR",
}
`;

exports[`createScene 2`] = `
Object {
  "first": [Function],
  "fourth": [Function],
  "notFound": [Function],
  "plain": [Function],
  "second": [Function],
  "third": [Function],
}
`;

exports[`createScene 3`] = `
"const { types, actions } = createScene(routes, { scene: 'SCENE' })

const { 
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
} = types

const { 
	first,
	second,
	third,
	fourth,
	plain,
	notFound
} = actions

export {
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
}

export {
	first,
	second,
	third,
	fourth,
	plain,
	notFound
}"
`;

exports[`double createScene 1`] = `
Object {
  "FIRST": "double/scene/FIRST",
  "FIRST_COMPLETE": "double/scene/FIRST_COMPLETE",
  "FIRST_ERROR": "double/scene/FIRST_ERROR",
  "FOURTH": "double/scene/FOURTH",
  "FOURTH_COMPLETE": "double/scene/FOURTH_COMPLETE",
  "FOURTH_ERROR": "double/scene/FOURTH_ERROR",
  "NOT_FOUND": "double/scene/@@rudy/NOT_FOUND",
  "NOT_FOUND_COMPLETE": "double/scene/@@rudy/NOT_FOUND_COMPLETE",
  "NOT_FOUND_ERROR": "double/scene/@@rudy/NOT_FOUND_ERROR",
  "PLAIN": "double/scene/PLAIN",
  "PLAIN_COMPLETE": "double/scene/PLAIN_COMPLETE",
  "PLAIN_ERROR": "double/scene/PLAIN_ERROR",
  "SECOND": "double/scene/SECOND",
  "SECOND_COMPLETE": "double/scene/SECOND_COMPLETE",
  "SECOND_ERROR": "double/scene/SECOND_ERROR",
  "THIRD": "double/scene/THIRD",
  "THIRD_COMPLETE": "double/scene/THIRD_COMPLETE",
  "THIRD_ERROR": "double/scene/THIRD_ERROR",
}
`;

exports[`double createScene 2`] = `
Object {
  "first": [Function],
  "fourth": [Function],
  "notFound": [Function],
  "plain": [Function],
  "second": [Function],
  "third": [Function],
}
`;

exports[`double createScene 3`] = `
Object {
  "@@rudy/NOT_FOUND": Object {
    "path": "/not-found",
    "type": "@@rudy/NOT_FOUND",
  },
  "double/scene/@@rudy/NOT_FOUND": Object {
    "path": "/not-found-foo",
    "scene": "double/scene",
    "type": "double/scene/@@rudy/NOT_FOUND",
  },
  "double/scene/FIRST": Object {
    "path": "/first",
    "scene": "double/scene",
    "type": "double/scene/FIRST",
  },
  "double/scene/FOURTH": Object {
    "action": [Function],
    "onComplete": [Function],
    "path": "/fourth",
    "scene": "double/scene",
    "thunk": [Function],
    "type": "double/scene/FOURTH",
  },
  "double/scene/PLAIN": Object {
    "action": [Function],
    "scene": "double/scene",
    "type": "double/scene/PLAIN",
  },
  "double/scene/SECOND": Object {
    "error": [Function],
    "path": "/second",
    "scene": "double/scene",
    "type": "double/scene/SECOND",
  },
  "double/scene/THIRD": Object {
    "action": Array [
      "",
      "customCreator",
    ],
    "customCreator": [Function],
    "path": "/third",
    "scene": "double/scene",
    "thunk": [Function],
    "type": "double/scene/THIRD",
  },
}
`;

exports[`double createScene 4`] = `
"const { types, actions } = createScene(routes, { scene: 'double/scene' })

const { 
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
} = types

const { 
	first,
	second,
	third,
	fourth,
	plain,
	notFound
} = actions

export {
	FIRST,
	FIRST_COMPLETE,
	FIRST_ERROR,
	SECOND,
	SECOND_COMPLETE,
	SECOND_ERROR,
	THIRD,
	THIRD_COMPLETE,
	THIRD_ERROR,
	FOURTH,
	FOURTH_COMPLETE,
	FOURTH_ERROR,
	PLAIN,
	PLAIN_COMPLETE,
	PLAIN_ERROR,
	NOT_FOUND,
	NOT_FOUND_COMPLETE,
	NOT_FOUND_ERROR
}

export {
	first,
	second,
	third,
	fourth,
	plain,
	notFound
}"
`;
